documentation/milestones/individual_milestones/developer_tools.md

```diff
@@ -0,0 +1,221 @@
+- [ ] **Automated testing framework** is live and meets the following conditions:
+  - [ ] At least 1 unit test has been added to the codebase
+  - [ ] At least 1 integration test has been added to the codebase
+  - [ ] Unit tests are run on every push to the remote GitHub repository
+  - [ ] Integration tests are run at least once before merging new code into `main`
+  - [ ] Code which fails any of the unit or integration tests will be blocked from merging into `main`
+  - [ ] A report on the percentage of code covered by tests is available after every test run
+  - [ ] Code whose test coverage falls below a certain threshold will be blocked from merging into `main`
+- [ ] **Code quality checks** are live and meet the following conditions:
+  - [ ] All checks are run on every push to the remote GitHub repository
+  - [ ] The most important checks are run on every local commit
+  - [ ] Code which fails any of these checks will be blocked from merging into `main`
+- [ ] **Security checks** are live and meet the following conditions:
+  - [ ] At least 1 (test) security vulnerability report has been submitted
+  - [ ] Maintainers are notified within 1 hour of a vulnerability being reported within the grants API codebase
+  - [ ] Maintainers are notified within 72 hours of a vulnerability being reported on an upstream dependency
+  - [ ] Security checks are running on every push to the remote GitHub repository
+  - [ ] Code which fails any of these security checks is blocked from merging into `main`
+- [ ] **Dependency management** is live and meets the following conditions:
+  - [ ] At least 2 upstream dependencies have been added to the project
+  - [ ] Compatible versions of upstream dependencies are automatically detected when a new dependency is added
+  - [ ] Maintainers are notified when a new minor or major version of an upstream dependency is available
+- [ ] **Config & secrets management** is live and meets the following conditions:
+  - [ ] At least 1 configuration variable has been added to the project
+  - [ ] This config variable has different values when it is running in different environments (e.g. dev, staging, and prod)
+  - [ ] At least 2 internal developers can access a shared value of this variable locally
+  - [ ] Developers can change the value of this config variable in their local environment without changing it for other developers
+  - [ ] An integration test has been created to validate that this variable is injected at runtime
+
+### Proposed metrics for measuring goals/value/definition of done
+
+- Test framework metrics
+  - Percentage of code covered by unit tests
+  - Percentage of code covered by integration tests
+  - Total runtime of all unit tests
+  - Total runtime of all integration tests
+- Code quality checks metric: Total runtime of all code quality checks
+- Security checks metrics
+  - Total runtime of all security checks
+  - Number of security vulnerabilities not caught until production
```

Comment:
These three ("Number of...") are good metrics, but I'm thinking about whether there would be a way to automate their calculation and reporting. It seems a little tricky to do so?

We don't want to be proposing metrics that are hard or too time-consuming to track.

Reply:
Good call! I simplified and cleaned up all of metrics to keep them a bit more focused in [this commit](https://github.com/HHS/grants-equity/pull/24/commits/bc3a5e6c2c4ab28b7c9bede6b3dd5c0a735db4d2)

```
- Total runtime of all CI/CD checks
- Number of bug tickets opened in the last 30 days
- Number of days it takes to close a bug ticket
- Percentage of code covered by unit tests
- Percentage of code covered by integration tests
```